x-pack/plugin/data-streams/src/main/java/org/elasticsearch/xpack/datastreams/mapper/DataTierFieldMapper.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.datastreams.mapper;
+
+
+import org.apache.lucene.search.MatchAllDocsQuery;
+import org.apache.lucene.search.Query;
+import org.elasticsearch.cluster.node.DiscoveryNodeRole;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.regex.Regex;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.mapper.ConstantFieldType;
+import org.elasticsearch.index.mapper.MetadataFieldMapper;
+import org.elasticsearch.index.mapper.ValueFetcher;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.node.NodeRoleSettings;
+import org.elasticsearch.search.lookup.SearchLookup;
+import org.elasticsearch.xpack.core.DataTier;
+
+import java.util.Collections;
+import java.util.function.Supplier;
+
+public class DataTierFieldMapper extends MetadataFieldMapper {
+
+    public static final String NAME = "_tier";
+
+    public static final String CONTENT_TYPE = "_tier";
+
+    public static final TypeParser PARSER = new FixedTypeParser(c -> new DataTierFieldMapper());
+
+    static final class DataTierFieldType extends ConstantFieldType {
+
+        static final DataTierFieldType INSTANCE = new DataTierFieldType();
+
+        private DataTierFieldType() {
+            super(NAME, Collections.emptyMap());
+        }
+
+        @Override
+        public String typeName() {
+            return CONTENT_TYPE;
+        }
+                
+
+        @Override
+        protected boolean matches(String pattern, boolean caseInsensitive, SearchExecutionContext context) {
+            if (caseInsensitive) {
+                pattern = Strings.toLowercaseAscii(pattern);
+            }
+            //hot warm etc tier name settings are prefixed with "data_"
+            final String normalised_pattern = "data_" + pattern;
+            Settings nodeSettings = context.getIndexSettings().getNodeSettings();
+            // Check any explicit settings picked by node
+            if (DataTier.matchesNodeRole(nodeSettings, role -> roleMatch(normalised_pattern, role))) {
```

Comment:
This check is not needed. The line below will return explicit settings if they exist and the defaults one otherwise. So you should just rely on what ` NodeRoleSettings.NODE_ROLES_SETTING.get(nodeSettings)` returns.

Reply:
>you should just rely on what NodeRoleSettings.NODE_ROLES_SETTING.get(nodeSettings) returns.

I wanted to keep DataTier dealing with simple String types rather than introducing a dependency on DiscoveryNodeRole. I can add it though if that's the lesser of the two evils